watertap/property_models/anaerobic_digestion/modified_adm1_reactions.py

```diff
@@ -0,0 +1,2564 @@
+                "S_bu"
+            ] / (self.params.K_a_bu + self.S_H)
+
+        self.concentration_of_bu = pyo.Constraint(
+            rule=concentration_of_bu_rule,
+            doc="constraint concentration of bu-",
+        )
+
+        def concentration_of_pro_rule(self):
+            return self.conc_mass_pro == self.params.K_a_pro * self.conc_mass_comp_ref[
+                "S_pro"
+            ] / (self.params.K_a_pro + self.S_H)
+
+        self.concentration_of_pro = pyo.Constraint(
+            rule=concentration_of_pro_rule,
+            doc="constraint concentration of pro-",
+        )
+
+        def concentration_of_ac_rule(self):
+            return self.conc_mass_ac == self.params.K_a_ac * self.conc_mass_comp_ref[
+                "S_ac"
+            ] / (self.params.K_a_ac + self.S_H)
+
+        self.concentration_of_ac = pyo.Constraint(
+            rule=concentration_of_ac_rule,
+            doc="constraint concentration of ac-",
+        )
+
+        def concentration_of_hco3_rule(self):
+            return self.conc_mol_hco3 == self.K_a_co2 * (
+                self.conc_mass_comp_ref["S_IC"] / (12 * pyo.units.kg / pyo.units.kmol)
+            ) / (self.K_a_co2 + self.S_H)
+
+        self.concentration_of_hco3 = pyo.Constraint(
+            rule=concentration_of_hco3_rule,
+            doc="constraint concentration of hco3",
+        )
+
+        def concentration_of_nh3_rule(self):
+            return self.conc_mol_nh3 == self.K_a_IN * (
+                self.conc_mass_comp_ref["S_IN"] / (14 * pyo.units.kg / pyo.units.kmol)
+            ) / (self.K_a_IN + self.S_H)
+
+        self.concentration_of_nh3 = pyo.Constraint(
+            rule=concentration_of_nh3_rule,
+            doc="constraint concentration of nh3",
+        )
+
+        # TO DO: use correct conversion number
+        def concentration_of_co2_rule(self):
+            return (
+                self.conc_mol_co2
+                == self.conc_mass_comp_ref["S_IC"]
+                / (12 * pyo.units.kg / pyo.units.kmol)
+                - self.conc_mol_hco3
+            )
+
+        self.concentration_of_co2 = pyo.Constraint(
+            rule=concentration_of_co2_rule,
+            doc="constraint concentration of co2",
+        )
+
+        def concentration_of_nh4_rule(self):
+            return (
+                self.conc_mol_nh4
+                == self.conc_mass_comp_ref["S_IN"]
+                / (14 * pyo.units.kg / pyo.units.kmol)
+                - self.conc_mol_nh3
+            )
+
+        self.concentration_of_nh4 = pyo.Constraint(
+            rule=concentration_of_nh4_rule,
+            doc="constraint concentration of pro-",
+        )
+
+        def S_OH_rule(self):
+            return self.S_OH == self.KW / self.S_H
+
+        self.S_OH_cons = pyo.Constraint(
+            rule=S_OH_rule,
+            doc="constraint concentration of OH",
+        )
+
+        def S_H_rule(self):
+            return (
+                self.state_ref.cations
```

Comment:
So, now that potassium and magnesium are tracked components, shouldn't they be incorporated into this constraint? Perhaps we need to double-check the reference to see how these components are handled. Overall, I would think that they need to be tied to `cations` and `anions`, at least indirectly (i.e., included in this constraint) if not directly (e.g., add constraint that relates potassium and magnesium concentrations to the total cation and anion concentrations).

Reply:
@MarcusHolly How were K and Mg handled in the literature? I would assume they'd account for those species as I suggested and as you have now implemented; I just want a sanity check on how well our implementation reflects the Flores et al. reference.

Reply:
> > I've tried re-running the MacOS setup test, but it still fails. Is it fine if this test is just skipped?
> 
> So, I think it depends on what is failing in the test. Originally, it was just a matter of absolute tolerance. However, after latest changes, it appears to be more of a scaling issue (I suspect) that now throws off the AD unit's initialization. It'd be better to try to resolve that before we resort to skipping. That said, it'd be helpful for a Mac user to tinker with this. @bknueven could you take a look at this? I am sure you will find that there is substantial room for improvement in scaling and initialization for these models, but for now, we're just looking to hit the _bare minimum requirement_ to get an initial implementation passing tests and merged. Following that, we'd have subsequent PRs to address scaling and initialization.

The other approach we can take is to skip the Mac test for now, merge this PR, get the associated translator block PR approved and merged, then come back and refine scaling, etc. 

However, please see my inquiry about additional species inclusion (struvite and calcium phosphate).

Reply:
I think your question is essentially where is this equation coming from? I just modified the equation presented in the reference for the regular adm1 reaction file: 
![image](https://user-images.githubusercontent.com/96305519/228541161-da4c210b-e48d-4f38-bf4a-898113bb3781.png)
Our reference (from what I can tell) does not spell out how K and Mg are handled. If it does, it'd probably be in one of the MATLAB files.

Reply:
> > > I've tried re-running the MacOS setup test, but it still fails. Is it fine if this test is just skipped?
> > 
> > 
> > So, I think it depends on what is failing in the test. Originally, it was just a matter of absolute tolerance. However, after latest changes, it appears to be more of a scaling issue (I suspect) that now throws off the AD unit's initialization. It'd be better to try to resolve that before we resort to skipping. That said, it'd be helpful for a Mac user to tinker with this. @bknueven could you take a look at this? I am sure you will find that there is substantial room for improvement in scaling and initialization for these models, but for now, we're just looking to hit the _bare minimum requirement_ to get an initial implementation passing tests and merged. Following that, we'd have subsequent PRs to address scaling and initialization.
> 
> The other approach we can take is to skip the Mac test for now, merge this PR, get the associated translator block PR approved and merged, then come back and refine scaling, etc.
> 
> However, please see my inquiry about additional species inclusion (struvite and calcium phosphate).

This is probably something we'll talk about more during today's meeting, but there is a complication with the Translator block PR. Our current implementation of ASM2d does not include S_IC, S_K, or S_Mg terms. So we will either need to update our current ASM2d implementation or add a "modified_asm2d". I'd probably lean towards the latter. I'm currently looking for literature sources on this, but my first instinct is to use the ASM GitHub created by the same authors: https://github.com/wwtmodels/Activated-Sludge-Models. My immediate concern with this is that everything appears to be inside MATLAB files this time (there's no Excel with results)